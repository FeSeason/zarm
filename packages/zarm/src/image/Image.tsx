import React from 'react';
import classnames from 'classnames';
import { ConfigContext } from '../n-config-provider';
import useImage, { IMAGE_STATUS } from './useImage';

export interface ImageCssVars {
  '--default-background-color'?: React.CSSProperties['color'];
  '--default-text-color'?: React.CSSProperties['color'];
  '--default-font-size'?: React.CSSProperties['fontSize'];
}

export interface ImageProps extends Omit<React.ImgHTMLAttributes<HTMLImageElement>, 'placeholder'> {
  placeholder?: React.ReactNode | boolean;
  fallback?: React.ReactNode | boolean;
}

const Image = React.forwardRef<HTMLImageElement, ImageProps>((props, ref) => {
  const {
    src,
    placeholder,
    fallback,
    alt,
    onLoad,
    onError,
    className,
    width,
    height,
    style,
    ...restProps
  } = props;

  const { prefixCls: globalPrefixCls } = React.useContext(ConfigContext);
  const prefixCls = `${globalPrefixCls}-image`;
  const cls = classnames(prefixCls, className);

  const imgRef = ref || React.createRef<HTMLImageElement>();
  const status = useImage({ src, onLoad, onError });

  const renderImageStatus = {
    [IMAGE_STATUS.LOADED]: () => (
      <img className={`${prefixCls}__img`} src={src} alt={alt} ref={imgRef} {...restProps} />
    ),
    [IMAGE_STATUS.FAILED]: () => {
      const renderFallback = (text) => <div className={`${prefixCls}__fallback`}>{text}</div>;

      if (fallback) {
        return React.isValidElement(fallback) ? fallback : renderFallback(fallback);
      }

      return (
        <img className={`${prefixCls}__img`} src={src} alt={alt} ref={imgRef} {...restProps} />
      );
    },
    [IMAGE_STATUS.LOADING]: () => {
      if (placeholder) {
        return React.isValidElement(placeholder) ? (
          placeholder
        ) : (
          <div className={`${prefixCls}__loading`}>{placeholder}</div>
        );
      }

      return (
        <img className={`${prefixCls}__img`} src={src} alt={alt} ref={imgRef} {...restProps} />
      );
    },
  };

  return (
    <div className={cls} style={{ width, height, ...style }}>
      {renderImageStatus[status]()}
    </div>
  );
});

Image.displayName = 'Image';
Image.defaultProps = {
  width: '100%',
  height: '100%',
  placeholder: true,
  fallback: true,
};

export default Image;
