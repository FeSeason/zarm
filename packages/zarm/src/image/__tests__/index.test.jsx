import React from 'react';
import toJson from 'enzyme-to-json';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import Image from '../index';

const src =
  'src="https://camo.githubusercontent.com/f5847256d81e5f8c31aee6554f749baf64654a131fed0fca987bd39e023a690f/68747470733a2f2f7a61726d2e64657369676e2f696d616765732f6c6f676f2e31613663666333302e737667"';

describe('Image', () => {
  it('renders correctly', () => {
    const wrapper = render(<Image src={src} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('width', () => {
    const wrapper = render(<Image src={src} width={40} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('height', () => {
    const wrapper = render(<Image src={src} height={40} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('placeholder', () => {
    const wrapper = render(<Image src={src} placeholder="loading" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('alt', () => {
    const wrapper = render(<Image src={src} alt="image" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('fallback', () => {
    const wrapper = render(<Image src={src} fallback={<span>图片不见啦</span>} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  // it('onLoad', () => {
  //   const onLoad = jest.fn();
  //   render(<Image src={src} onLoad={onLoad} />);
  //   expect(onLoad).toHaveBeenCalled();
  // });

  it('onError', async () => {
    const onError = jest.fn();
    const dom = render(
      <Image src="xxx" alt="test" fallback={false} placeholder={false} onError={onError} />,
    );
    const img = dom.querySelector('img');
    console.log(img);
    fireEvent.error(dom.querySelector('img'));
    jest.setTimeout(5000);
    await waitFor(() => expect(onError).toHaveBeenCalled());
  });
});
