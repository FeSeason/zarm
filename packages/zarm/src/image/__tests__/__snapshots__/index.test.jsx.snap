// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image alt 1`] = `undefined`;

exports[`Image fallback 1`] = `undefined`;

exports[`Image height 1`] = `undefined`;

exports[`Image placeholder 1`] = `undefined`;

exports[`Image render correctly 1`] = `undefined`;

exports[`Image width 1`] = `undefined`;
